### PR TITLE
Detect presence of extern char **environ

### LINKED ISSUE #952: Undefined symbols: "_environ"
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Mac OS X 10.6.8 i386 32
| File format of the file you reverse (mandatory)      | n/a
| Architecture/bits of the file (mandatory)            | n/a
| `rizin -v` full output, **not truncated** (mandatory)  | n/a

### Expected behavior

Successful build from source

### Actual behavior

Build failure:

```
Undefined symbols for architecture i386:
  "_environ", referenced from:
      _rz_subprocess_start in subprocess.c.o
     (maybe you meant: _rz_sys_get_environ, _rz_sys_set_environ )
ld: symbol(s) not found for architecture i386
```

We know that old versions of Mac OS X don't have `_environ` and you appear to already have code in librz/util/sys.c to deal with that:

```c
RZ_API char **rz_sys_get_environ(void) {
#if __APPLE__ && !HAVE_ENVIRON
	env = *_NSGetEnviron();
#else
	env = environ;
#endif
	// return environ if available??
	if (!env) {
		env = rz_lib_dl_sym(NULL, "environ");
	}
	return env;
}
```

But something isn't working.

### Additional Logs, screenshots, source code,  configuration dump, ...

Build log: [log.txt.zip](https://github.com/rizinorg/rizin/files/6246872/log.txt.zip)

### PR DESCRIPTION
**Pull request by XVilka**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Detecting if the platform has `extern char **environ` variable.

**Test plan**

All green, look for `HAVE_ENVIRON` in the logs.

**Closing issues**

Closes #952

### CODE REVIEW

**Comment by XVilka** [comment]
Strange, seems to break subprocess unit test:
```
test_env ERR
[XX] Fail at line 63: YOUVAR env var should be passed and printed on stdout: expected Hello Rizin Project
, got Hello World
.
```

File: librz/util/subprocess.c [hunk 1]
```diff
@@ -764,7 +764,7 @@ RZ_API RzSubprocess *rz_subprocess_start_opt(RzSubprocessOpt *opt) {
 		}
 
 		// Use the previously created environment
-		environ = child_env;
+		rz_sys_set_environ(child_env);
```
Comment (ret2libc) [inline librz/util/subprocess.c:767]: I think the problem is that this function does not really set `environ`, but it just sets a `static char **env` var defined in sys.c.

Comment (XVilka) [reply librz/util/subprocess.c:767]: Good point. I fixed that function, please check https://github.com/rizinorg/rizin/pull/953/commits/ecb5fd364de2d0642ab1b264e8d45c92a6255325

File: librz/util/sys.c [hunk 2]
```diff
@@ -1252,7 +1252,7 @@ RZ_API char *rz_sys_pid_to_path(int pid) {
 RZ_API void rz_sys_env_init(void) {
 	char **envp = rz_sys_get_environ();
 	if (envp) {
-		rz_sys_set_environ(envp);
```
Comment (XVilka) [inline librz/util/sys.c:1255]: This function isn't used anywhere else.

File: meson.build [hunk 3]
```diff
@@ -331,6 +331,11 @@ if not cc.has_function('clock_gettime', prefix: '#include <time.h>') and cc.has_
   lrt = cc.find_library('rt', required: true, static: is_static_build)
 endif
 
+code = '#include <unistd.h>\nextern char **environ;\nint main() { char **env = environ; }'
```
Comment (ret2libc) [inline meson.build:334]: there are a bunch of places in the code that try to define HAVE_ENVIRON. Can't we remove them and just use  this?

File: meson.build [hunk 4]
```diff
@@ -331,6 +331,11 @@ if not cc.has_function('clock_gettime', prefix: '#include <time.h>') and cc.has_
...
+have_environ = cc.links(code, name: 'have extern char **environ')
+userconf.set10('HAVE_ENVIRON', have_environ)
+message('HAVE_ENVIRON: @0@'.format(have_environ))
```
Comment (ret2libc) [inline meson.build:337]: this is useless as long as you don't put `HAVE_ENVIRON` in rz_userconf.h.in and you don't set it like all the others vars.

**Review by ret2libc — CHANGES_REQUESTED** [review]

**Review by ret2libc — APPROVED** [review]

**Pull request merged**